lib/connector/OAuthConnector.js

```diff
@@ -182,52 +253,155 @@ class OAuthConnector {
+                    if (!isNaN(userContext.vendorToken.expires_in)) {
+                        userContext.vendorToken.expires_at = Date.now() + +userContext.vendorToken.expires_in * 1000;
+                    }
+                    userContext.vendorUserProfile = await this.getUserProfile(userContext.vendorToken);
+                    userContext.status = 'authenticated';
+                    userContext.refreshErrorCount = 0;
+                    await this.saveUser(fusebitContext, userContext);
+                    return userContext.vendorToken;
+                } catch (e) {
+                    if (userContext.refreshErrorCount > this.refreshErrorLimit) {
+                        Sdk.debug('REFRESH TOKEN ERROR, DELETING USER', e);
+                        await this.deleteUser(fusebitContext, userContext.vendorUserId);
+                        throw new Error(
+                            `Error refreshing access token. Maximum number of attempts exceeded, user has been deleted: ${e.message}`
+                        );
+                    } else {
+                        userContext.refreshErrorCount = (userContext.refreshErrorCount || 0) + 1;
+                        userContext.status = 'refresh_error';
+                        await this.saveUser(fusebitContext, userContext);
+                        throw new Error(
+                            `Error refreshing access token, attempt ${userContext.refreshErrorCount} out of ${this.refreshErrorLimit}: ${e.message}`
+                        );
+                    }
+                }
+            }
+        };
+
+        const waitForRefreshedAccessToken = async (count, backoff) => {
+            Sdk.debug('WAITING FOR ACCESS TOKEN TO BE REFRESHED FOR USER', userContext.vendorUserId, 'ATTEMPTS LEFT', count);
+            if (!count > 0) {
+                throw new Error(`Error refreshing access token. Waiting for the access token to be refreshed exceeded the maximum time`);
+            }
+            return new Promise((resolve, reject) => {
+                setTimeout(async () => {
+                    try {
+                        userContext = await this.getUser(fusebitContext, userContext.vendorUserId);
+                        if (!userContext || userContext.status === 'refresh_error') {
+                            throw new Error(`Concurrent access token refresh operation failed`);
+                        }
+                    } catch (e) {
+                        return reject(new Error(`Error waiting for access token refresh: ${e.message}`));
+                    }
+                    if (userContext.status === 'authenticated') {
+                        return resolve(userContext.vendorToken);
+                    } else {
+                        let result;
+                        try {
+                            result = await this.waitForRefreshedAccessToken(count - 1, Math.floor(backoff * this.refreshBackoffIncrement));
```

Comment:
Consider flattening to not deepen the stack on each recursion.

Reply:
The setTimeout around it breaks the recursion.